### PR TITLE
Make HA test helpers deal with sets.

### PR DESCRIPTION
**Pull request by mattmoor**
The previous helpers dealt with singleton leaders, change them to accommodate bucketing.

I am starting to experiment with these downstream here: https://github.com/knative/serving/pull/8588

### CODE REVIEW

File: test/ha/ha.go [hunk 1]
```diff
@@ -23,12 +23,79 @@ import (
...
+		pod := strings.Split(*lease.Spec.HolderIdentity, "_")[0]
+
+		// Deconstruct the pod name and look for the deployment.  This won't work for very long deployment names.
+		parts := strings.Split(pod, "-")
+		if len(parts) < 3 {
+			continue
+		}
+		if strings.Join(parts[:len(parts)-2], "-") != deploymentName {
+			continue
+		}
```
Comment (vagababov) [inline test/ha/ha.go:52]: I wonder if we can use one split/find call?

File: test/ha/ha.go [hunk 2]
```diff
@@ -23,12 +23,79 @@ import (
...
+	return wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		currentLeaders, err := GetLeaders(client, deploymentName, namespace)
+		if err != nil {
+			return false, err
+		}
+		if previousLeaders.HasAny(currentLeaders...) {
+			// TODO(mattmoor): Log overlap?
+			return false, nil
+		}
+		return true, nil
```
Comment (vagababov) [inline test/ha/ha.go:94]: `return !previousLeaders.HasAny(currentLeaders...), nil`

File: test/ha/ha.go (hunk 1 above)
Comment (mattmoor) [reply test/ha/ha.go:52]: 🤷

File: test/ha/ha.go (hunk 2 above)
Comment (mattmoor) [reply test/ha/ha.go]: I combined the methods to be closer to the style in use before, so this is a bit impractical now.

File: test/ha/ha.go [hunk 3]
```diff
@@ -23,12 +23,79 @@ import (
...
+	leases, err := client.Kube.CoordinationV1().Leases(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting leases for deployment %q: %w", deploymentName, err)
+	}
+	var pods []string
+	for _, lease := range leases.Items {
+		if lease.Spec.HolderIdentity == nil {
+			continue
+		}
+		pod := strings.Split(*lease.Spec.HolderIdentity, "_")[0]
```
[the remark this answers is not in the document]
Comment (mattmoor) [reply test/ha/ha.go]: done

**Comment by mattmoor** [comment]
I also plumbed through `t *testing.T` so I could `t.Log` (it was useful for debugging downstream)

**Comment by mattmoor** [comment]
I'll remove the `/hold` on this once the downstream PR passes, but I think the problem I've been chasing is: https://github.com/knative/serving/pull/8588/files#diff-0bff8fdbdcedf4cfbdcab24558df0df2R136-R138

**Comment by mattmoor** [comment]
/unhold

Tests pass.  I'd like to land this to that I can start moving downstream off of the old method and onto this. 🙏

File: test/ha/ha.go (hunk 1 above)
Comment (vagababov) [reply test/ha/ha.go:53]: 😢

**Pull request merged**